oat-sa/extension-tao-itemqti-pci
# feat: migrate to inline web worker

Related to: https://oat-sa.atlassian.net/browse/TR-3907

## Technical Acceptance criteria
AC1. The runtime of the Audio Recording PCI IMS doesn’t load external Web Worker files

## Technical consideration

The Web Worker of the PCI must be converted to Inline Web Workers, by loading the worker content as text, using the `require.js text! plugin, ie define(['text!path/to/workerFile.js'], function(workerContent) { `so it can be bundled with the PCI 

The web workers should be instantiated through blob URL (this approach has been tested cross browsers): 

```
const workerContentBlob = new Blob(['...content of the worker as text, ie. workerContent'],  { type: 'text/javascript' });
const workerContentUrl = URL.createObjectURL(workerContent);
const worker = worker = new Worker(workerContentUrl); 
```
Object URLs must be revoked when destroying the PCI to prevent memory leaks see [URL.revokeObjectURL() - Web APIs | MDN](https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL)

## How to test:
to run web worker you need set Uncompressed in settings 
<img width="206" alt="image" src="https://user-images.githubusercontent.com/25976342/164206133-8e480a15-343b-41c1-b6f3-d68ca5d90b8d.png">

1. create new item with audio PCI with Uncompressed recording 
2. create test, then delivery
3. run assessment and check interaction
4. login as admin and check in Results audio file and Review item
